pkg/storage/stores/index/labelvolume/labelvolume.go

```diff
@@ -45,53 +63,92 @@ func (acc *Accumulator) Volumes() *logproto.LabelVolumeResponse {
+			}
 
-		for _, v := range res.Volumes {
-			if _, ok := mergedVolumes[v.Name]; !ok {
-				mergedVolumes[v.Name] = make(map[string]uint64)
+			if _, ok := mergedVolumes[vol.Timestamp][name][value]; !ok {
+				mergedVolumes[vol.Timestamp][name][value] = vol.Value
+			} else {
+				mergedVolumes[vol.Timestamp][name][value] += vol.Value
 			}
-			mergedVolumes[v.Name][v.Value] += v.GetVolume()
 		}
 	}
 
+	return MapToLabelVolumeResponse(mergedVolumes, int(acc.limit))
+}
+
+func Merge(responses []*logproto.LabelVolumeResponse, limit int32) *logproto.LabelVolumeResponse {
+	mergedVolumes := MergeVolumes(responses...)
 	return MapToLabelVolumeResponse(mergedVolumes, int(limit))
 }
 
-func MapToLabelVolumeResponse(mergedVolumes map[string]map[string]uint64, limit int) *logproto.LabelVolumeResponse {
-	volumes := make([]logproto.LabelVolume, 0, len(mergedVolumes))
-	for name, v := range mergedVolumes {
-		for value, volume := range v {
+func MapToLabelVolumeResponse(mergedVolumes map[int64]map[string]map[string]uint64, limit int) *logproto.LabelVolumeResponse {
+  // Aggregate volumes into single value per label/value pair
+  // adopting the latest timestamp seen
+	acc := make(map[string]map[string]*volume)
+	for timestamp, series := range mergedVolumes {
+		for name, vs := range series {
+			for value, v := range vs {
+				if _, ok := acc[name]; !ok {
+					acc[name] = make(map[string]*volume)
+				}
+
+				if vol, ok := acc[name][value]; !ok {
+					acc[name][value] = &volume{
+						Timestamp: timestamp,
+						Value:     v,
+					}
+				} else {
+					if vol.Timestamp < timestamp {
+						vol.Timestamp = timestamp
+					}
+
+					vol.Value += v
+				}
+			}
+		}
+	}
+
+  // Convert aggregation into slice of logproto.LabelVolume
+	volumes := make([]logproto.LabelVolume, 0, len(mergedVolumes)*len(mergedVolumes[0]))
+	for name, vs := range acc {
+		for value, v := range vs {
 			volumes = append(volumes, logproto.LabelVolume{
-				Name:   name,
-				Value:  value,
-				Volume: volume,
+				Name:      name,
+				Value:     value,
+				Volume:    v.Value,
+				Timestamp: v.Timestamp,
 			})
 		}
 	}
 
+  // Sort the LabelVolume response by timestamp, volume, name, value
+  // to ensure consistency in responses
 	sort.Slice(volumes, func(i, j int) bool {
-		if volumes[i].Volume == volumes[j].Volume {
-			if volumes[i].Name == volumes[j].Name {
-				return volumes[i].Value < volumes[j].Value
+		if volumes[i].Timestamp == volumes[j].Timestamp {
+			if volumes[i].Volume == volumes[j].Volume {
+				if volumes[i].Name == volumes[j].Name {
+					return volumes[i].Value < volumes[j].Value
+				}
+
+				return volumes[i].Name < volumes[j].Name
 			}
 
-			return volumes[i].Name < volumes[j].Name
+			return volumes[i].Volume > volumes[j].Volume
 		}
-
-		return volumes[i].Volume > volumes[j].Volume
+		return volumes[i].Timestamp < volumes[j].Timestamp
 	})
 
-	if limit < len(volumes) {
+  // Apply the limit
+	if len(volumes) > limit {
```

Comment:
would be curious on your feedback here. based on the sort, this is going to limit to the first N by timestamp, whereas we probably want the last N, but then also the biggest. Maybe we should change the sort to be biggest timestamp first? Or maybe we don't care about the timestamp here?